packages/react-router-with-query/src/index.tsx

```diff
@@ -4,19 +4,23 @@ import {
   hashKey,
   hydrate,
 } from '@tanstack/react-query'
-import { type AnyRouterWithContext } from '@tanstack/react-router'
 import { Fragment } from 'react'
+import type { AnyRouter, AnyRouterWithContext } from '@tanstack/react-router'
 import type {
   QueryClient,
   QueryObserverResult,
   UseQueryOptions,
 } from '@tanstack/react-query'
 
-export function routerWithQueryClient<
-  T extends AnyRouterWithContext<{
-    queryClient: QueryClient
-  }>,
->(router: T, queryClient: QueryClient): T {
+export function routerWithQueryClient<TRouter extends AnyRouter>(
+  _router: TRouter extends AnyRouterWithContext<infer Ctx>
+    ? Ctx extends { queryClient: QueryClient }
+      ? TRouter
+      : 'Context type mismatch. Expected queryClient to be in context type.'
+    : 'Context type mismatch. Expected queryClient to be in context type.',
+  queryClient: QueryClient,
+): TRouter {
+  const router = _router as unknown as AnyRouter
```

Comment:
```suggestion
export function routerWithQueryClient<TRouter extends AnyRouter>(
  router: TRouter extends AnyRouterWithContext<infer Ctx>
    ? Ctx extends { queryClient: QueryClient }
      ? TRouter
      : never
    : never,
  queryClient: QueryClient,
): TRouter {
```

Reply:
when using `never`, you don't have to do any type assertions

Reply:
unfortunately it appears I have to in order to support all the ts versions:
![CleanShot 2024-07-23 at 10 04 29](https://github.com/user-attachments/assets/801b8912-a5cb-41b6-b11b-8cf7ba39fc2a)

https://github.com/TanStack/router/pull/2007#discussion_r1686586556

Reply:
@chorobin can you please have a look at this?

Reply:
I feel like I’ve run into this before too. Very odd.

Reply:
Fairly sure this is due to invariance. If a type parameter is in a `in` and an `out` position then a router with a different context type is not assignable to `AnyRouterWithContext`

Reply:
Why not
```typescript
export function routerWithQueryClient<TRouter extends AnyRouter>(
  router: TRouter['options']['context'] extends { queryClient: QueryClient }
    ? TRouter
    : AnyRouterWithContext<{ queryClient: QueryClient }>,
  queryClient: QueryClient,
): TRouter
```

Reply:
Well, what do we do? We need this to move forward.

Reply:
I would say my above suggestion works. What do you think @schiller-manuel

Reply:
yes please go for https://github.com/TanStack/router/pull/2007#discussion_r1691905804

Reply:
Will see if it works for all versions

Reply:
![CleanShot 2024-07-26 at 23 38 33](https://github.com/user-attachments/assets/76262b96-e06b-408a-9176-b5a21c018920)
 Nope...